guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+so we can more easily pass it around.
+
+The compiler actually tells us what the fix is. Add `T: 'static`:
+
+```rust
+impl<T> Handler for Timeout<T>
+where
+    T: Handler + Clone + 'static,
+{
+    // ...
+}
+```
+
+It now compiles!
+
+Lets create a similar handler struct for adding a `Content-Type` header on the
+response:
+
+```rust
+#[derive(Clone)]
+struct JsonContentType<T> {
+    inner_handler: T,
+}
+
+impl<T> Handler for JsonContentType<T>
+where
+    T: Handler + Clone + 'static,
+{
+    type Future = Pin<Box<dyn Future<Output = Result<Response, Error>>>>;
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        let mut this = self.clone();
+
+        Box::pin(async move {
+            let mut response = this.inner_handler.call(request).await?;
+            response.set_header("Content-Type", "application/json");
+            Ok(response)
+        })
+    }
+}
+```
+
+Notice this follows a very similar pattern to `Timeout`.
+
+Next up we modify `Server::run` to accept our new `Handler` trait:
+
+```rust
+impl Server {
+    async fn run<T>(self, mut handler: T) -> Result<(), Error>
+    where
+        T: Handler,
+    {
+        loop {
+            let mut connection = receive_connection().await?;
+            let request = parse_http_request(&mut connection).await?;
+            // have to call `Handler::call` here
+            match handler.call(request).await {
+                Ok(response) => write_http_response(connection).await?,
+                Err(error) => handle_error_somehow(error),
+            }
+        }
+    }
+}
+```
+
+We can now compose our three handlers together and run our server like so:
+
+```rust
+JsonContentType {
+    inner_handler: Timeout {
+        inner_handler: RequestHandler,
+        duration: Duration::from_secs(30),
+    },
+}
+```
+
+And if we add some `new` methods to our types they become a little easier to
+compose:
+
+```rust
+let handler = RequestHandler;
+let handler = Timeout::new(handler, Duration::from_secs(30));
+let handler = JsonContentType::new(handler);
+
+// `handler` has type `JsonContentType<Timeout<RequestHandler>>`
+
+server.run(handler).await
+```
+
+This works quite well! We're now able to layer on additional functionality to
+`RequestHandler` without having to modify its implementation. In theory we could
+put our `JsonContentType` and `Timeout` handlers into a crate and release it as
+library on crates.io for other people to use!
+
+## Making `Handler` more flexible
+
+Our little `Handler` trait is working quite nicely but currently it only
+supports our `Request` and `Response` types. It would be nice if those where
+generic so users could use whatever type they want:
+
+```rust
+// Lets make the request type a generic type parameter.
+// This allows users to implement `Handler` multiple times for
+// the same type. Its not something we need but our users might.
+trait Handler<Request> {
```

Comment:
Might want to add a note why the `Request` is generic but `Response` is an associated type. This is because generally for traits if its a param for method the generic is much easier while a return type associated types work much better with how you compose things.

Reply:
Here's how I'd phrase this:

> We make `Request` a generic type parameter of the trait so that a given service can accept many different types of requests. We make `Response` an associated type because for any _given_ `Request` type, there can only be one (associated) `Response` type: the one the corresponding `call` returns!

Reply:
Love it!